### ISSUE TITLE
MethodError: no method matching Posit16(::BigFloat)

### ISSUE THREAD

**Issue by alexandrosstamatiou** [milankl/SoftPosit.jl#65]
Hello Milan,

I am learning to program in Julia and trying to use the package DifferentialEquations.jl (see https://github.com/SciML/DifferentialEquations.jl) in conjunction with SoftPosit.jl. 

According to the author of DifferentialEquations.jl, you can use any custom type as long as some basic operators are defined. However, I encounter the error message `MethodError: no method matching Posit16(::BigFloat)` when trying to solve a simple ODE problem using posits (as initial condition and time array). In fact, the same error occurs for me when defining a variable like `big_float = convert(BigFloat, 1.0)` and then constructing a 16-bit Posit with `Posit16(big_float)`. This suggests it has to do with SoftPosit.jl. The very same error message is thrown when using the alternative SigmoidNumbers.jl package. 

Do you have an idea what the issue might be here? 

Many thanks in advance and kind regards,

Alexandros

**Comment by milankl** [comment]
I agree that is weird, I don't know whether DiffEq uses some BigFloats initially (say weights of the some time integrator) to avoid any precision loss and then converts to the desired format?

**Comment by milankl** [comment]
@alexandrosstamatiou can you post the lines of your error message that contain the calling code from DiffEq?

**Comment by alexandrosstamatiou** [comment]
Sorry, I only got back to my desk this morning. Thank you very much for your help. I wil try this suggestion immediately. And thinking about the appearance of `BigFloat`, that's indeed a bit strange. So the problem I tried to solve is the most basic one from the DifferentialEquations.jl tutorials: 
Define the function:

`f(u,p,t) = 0.98u`

Then define the ODEProblem:

`using DifferentialEquations`
`u0 = 1.0   #define initial condition`
`tspan = (0.0,1.0)  #time span of interest`
`prob = ODEProblem(f,u0,tspan)`

Then we can solve using 'sol = solve(prob)' and it works fine, all with `Float64`, outputting the solution vectors.

Now I try the same with posits (using SigmoidNumbers.jl but I the problem will be the same for SoftPosit.jl):
`posit_u0 = Posit161.(u0)`
`posit_tspan = Posit161.(tspan)`
`posit_prob = ODEProblem(f, posit_u0, posit_tspan)`
Output check:
`ODEProblem with uType Posit{16,1} and tType Posit{16,1}. In-place: false
timespan: (Posit{16,1}(0x0000), Posit{16,1}(0x4000))
u0: Posit{16,1}(0x4000)`


Then running the command `posit_sol = solve(posit_prob)` results in the following error lines: 


MethodError: no method matching Posit{16,1}(::BigFloat)
Closest candidates are:
  (::Type{T})(::Real, ::RoundingMode) where T<:AbstractFloat at /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/base/rounding.jl:200
  (::Type{T})(::T) where T<:Number at /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/base/boot.jl:770
  (::Type{T})(::AbstractChar) where T<:Union{AbstractChar, Number} at /Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/base/char.jl:50
  ...

Stacktrace:
  [1] convert(#unused#::Type{Posit{16,1}}, x::BigFloat)
    @ Base ./number.jl:7
  [2] OrdinaryDiffEq.Vern9Tableau(T::Type, T2::Type)
    @ OrdinaryDiffEq ~/.julia/packages/OrdinaryDiffEq/ZBye7/src/tableaus/verner_tableaus.jl:3166
  [3] alg_cache(alg::Vern9, u::Posit{16,1}, rate_prototype::Posit{16,1}, #unused#::Type{Posit{16,1}}, #unused#::Type{Posit{16,1}}, #unused#::Type{Posit{16,1}}, uprev::Posit{16,1}, uprev2::Posit{16,1}, f::ODEFunction{false, typeof(f), LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, typeof(SciMLBase.DEFAULT_OBSERVED), Nothing}, t::Posit{16,1}, dt::Posit{16,1}, reltol::Posit{16,1}, p::SciMLBase.NullParameters, calck::Bool, #unused#::Val{false})
    @ OrdinaryDiffEq ~/.julia/packages/OrdinaryDiffEq/ZBye7/src/caches/verner_caches.jl:171
  [4] macro expansion
    @ ~/.julia/packages/OrdinaryDiffEq/ZBye7/src/caches/basic_caches.jl:0 [inlined]
  [5] __alg_cache(algs::Tuple{Vern9, Rodas5{0, false, Nothing, typeof(OrdinaryDiffEq.DEFAULT_PRECS), Val{:forward}, true, nothing}}, u::Posit{16,1}, rate_prototype::Posit{16,1}, #unused#::Type{Posit{16,1}}, #unused#::Type{Posit{16,1}}, #unused#::Type{Posit{16,1}}, uprev::Posit{16,1}, uprev2::Posit{16,1}, f::ODEFunction{false, typeof(f), LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, typeof(SciMLBase.DEFAULT_OBSERVED), Nothing}, t::Posit{16,1}, dt::Posit{16,1}, reltol::Posit{16,1}, p::SciMLBase.NullParameters, calck::Bool, #unused#::Val{false})
    @ OrdinaryDiffEq ~/.julia/packages/OrdinaryDiffEq/ZBye7/src/caches/basic_caches.jl:19
  [6] alg_cache(alg::CompositeAlgorithm{Tuple{Vern9, Rodas5{0, false, Nothing, typeof(OrdinaryDiffEq.DEFAULT_PRECS), Val{:forward}, true, nothing}}, OrdinaryDiffEq.AutoSwitchCache{Vern9, Rodas5{0, false, Nothing, typeof(OrdinaryDiffEq.DEFAULT_PRECS), Val{:forward}, true, nothing}, Rational{Int64}, Int64}}, u::Posit{16,1}, rate_prototype::Posit{16,1}, #unused#::Type{Posit{16,1}}, #unused#::Type{Posit{16,1}}, #unused#::Type{Posit{16,1}}, uprev::Posit{16,1}, uprev2::Posit{16,1}, f::ODEFunction{false, typeof(f), LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, typeof(SciMLBase.DEFAULT_OBSERVED), Nothing}, t::Posit{16,1}, dt::Posit{16,1}, reltol::Posit{16,1}, p::SciMLBase.NullParameters, calck::Bool, #unused#::Val{false})
    @ OrdinaryDiffEq ~/.julia/packages/OrdinaryDiffEq/ZBye7/src/caches/basic_caches.jl:14
  [7] __init(prob::ODEProblem{Posit{16,1}, Tuple{Posit{16,1}, Posit{16,1}}, false, SciMLBase.NullParameters, ODEFunction{false, typeof(f), LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, typeof(SciMLBase.DEFAULT_OBSERVED), Nothing}, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, SciMLBase.StandardODEProblem}, alg::CompositeAlgorithm{Tuple{Vern9, Rodas5{0, false, Nothing, typeof(OrdinaryDiffEq.DEFAULT_PRECS), Val{:forward}, true, nothing}}, AutoSwitch{Vern9, Rodas5{0, false, Nothing, typeof(OrdinaryDiffEq.DEFAULT_PRECS), Val{:forward}, true, nothing}, Rational{Int64}, Int64}}, timeseries_init::Tuple{}, ts_init::Tuple{}, ks_init::Tuple{}, recompile::Type{Val{true}}; saveat::Tuple{}, tstops::Tuple{}, d_discontinuities::Tuple{}, save_idxs::Nothing, save_everystep::Bool, save_on::Bool, save_start::Bool, save_end::Nothing, callback::Nothing, dense::Bool, calck::Bool, dt::Posit{16,1}, dtmin::Nothing, dtmax::Posit{16,1}, force_dtmin::Bool, adaptive::Bool, gamma::Rational{Int64}, abstol::Nothing, reltol::Nothing, qmin::Rational{Int64}, qmax::Int64, qsteady_min::Int64, qsteady_max::Int64, beta1::Nothing, beta2::Nothing, qoldinit::Rational{Int64}, controller::Nothing, fullnormalize::Bool, failfactor::Int64, maxiters::Int64, internalnorm::typeof(DiffEqBase.ODE_DEFAULT_NORM), internalopnorm::typeof(LinearAlgebra.opnorm), isoutofdomain::typeof(DiffEqBase.ODE_DEFAULT_ISOUTOFDOMAIN), unstable_check::typeof(DiffEqBase.ODE_DEFAULT_UNSTABLE_CHECK), verbose::Bool, timeseries_errors::Bool, dense_errors::Bool, advance_to_tstop::Bool, stop_at_next_tstop::Bool, initialize_save::Bool, progress::Bool, progress_steps::Int64, progress_name::String, progress_message::typeof(DiffEqBase.ODE_DEFAULT_PROG_MESSAGE), userdata::Nothing, allow_extrapolation::Bool, initialize_integrator::Bool, alias_u0::Bool, alias_du0::Bool, initializealg::OrdinaryDiffEq.DefaultInit, kwargs::Base.Pairs{Symbol, Bool, Tuple{Symbol, Symbol}, NamedTuple{(:default_set, :second_time), Tuple{Bool, Bool}}})
    @ OrdinaryDiffEq ~/.julia/packages/OrdinaryDiffEq/ZBye7/src/solve.jl:295
  [8] __solve(::ODEProblem{Posit{16,1}, Tuple{Posit{16,1}, Posit{16,1}}, false, SciMLBase.NullParameters, ODEFunction{false, typeof(f), LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, typeof(SciMLBase.DEFAULT_OBSERVED), Nothing}, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, SciMLBase.StandardODEProblem}, ::CompositeAlgorithm{Tuple{Vern9, Rodas5{0, false, Nothing, typeof(OrdinaryDiffEq.DEFAULT_PRECS), Val{:forward}, true, nothing}}, AutoSwitch{Vern9, Rodas5{0, false, Nothing, typeof(OrdinaryDiffEq.DEFAULT_PRECS), Val{:forward}, true, nothing}, Rational{Int64}, Int64}}; kwargs::Base.Pairs{Symbol, Bool, Tuple{Symbol, Symbol}, NamedTuple{(:default_set, :second_time), Tuple{Bool, Bool}}})
    @ OrdinaryDiffEq ~/.julia/packages/OrdinaryDiffEq/ZBye7/src/solve.jl:4
  [9] __solve(::ODEProblem{Posit{16,1}, Tuple{Posit{16,1}, Posit{16,1}}, false, SciMLBase.NullParameters, ODEFunction{false, typeof(f), LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, typeof(SciMLBase.DEFAULT_OBSERVED), Nothing}, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, SciMLBase.StandardODEProblem}, ::Nothing; default_set::Bool, kwargs::Base.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:second_time,), Tuple{Bool}}})
    @ DifferentialEquations ~/.julia/packages/DifferentialEquations/4jfQK/src/default_solve.jl:8
 [10] #__solve#59
    @ ~/.julia/packages/DiffEqBase/U3LtB/src/solve.jl:400 [inlined]
 [11] __solve
    @ ~/.julia/packages/DiffEqBase/U3LtB/src/solve.jl:387 [inlined]
 [12] #solve_call#39
    @ ~/.julia/packages/DiffEqBase/U3LtB/src/solve.jl:155 [inlined]
 [13] solve_call
    @ ~/.julia/packages/DiffEqBase/U3LtB/src/solve.jl:142 [inlined]
 [14] #solve_up#41
    @ ~/.julia/packages/DiffEqBase/U3LtB/src/solve.jl:185 [inlined]
 [15] solve_up
    @ ~/.julia/packages/DiffEqBase/U3LtB/src/solve.jl:173 [inlined]
 [16] #solve#40
    @ ~/.julia/packages/DiffEqBase/U3LtB/src/solve.jl:168 [inlined]
 [17] solve(::ODEProblem{Posit{16,1}, Tuple{Posit{16,1}, Posit{16,1}}, false, SciMLBase.NullParameters, ODEFunction{false, typeof(f), LinearAlgebra.UniformScaling{Bool}, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, Nothing, typeof(SciMLBase.DEFAULT_OBSERVED), Nothing}, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, SciMLBase.StandardODEProblem})
    @ DiffEqBase ~/.julia/packages/DiffEqBase/U3LtB/src/solve.jl:162
 [18] top-level scope
    @ In[17]:1
 [19] eval
    @ ./boot.jl:373 [inlined]
 [20] include_string(mapexpr::typeof(REPL.softscope), mod::Module, code::String, filename::String)
    @ Base ./loading.jl:1196

**Comment by alexandrosstamatiou** [comment]
Note: in the above I had to use SigmoidNumbers.jl as SoftPosit.jl won't precompile on my MacBook (M1 chip). I have successfully installed and used SoftPosit.jl on a Windows PC.

**Comment by milankl** [comment]
@giordano do you know why the precompilation of SoftPosit.jl on an apple M1 would fail although [softposit_jll](https://github.com/JuliaBinaryWrappers/SoftPosit_jll.jl) should build?

**Comment by milankl** [comment]
Regarding BigFloat, yes, as I assumed, the coefficients are first calculated as BigInts then converted

https://github.com/SciML/OrdinaryDiffEq.jl/blob/9470bf2833fce0b2d38323193127b4d49cf8874b/src/tableaus/verner_tableaus.jl#L3164-L3167

**Comment by alexandrosstamatiou** [comment]
Quick update: the suggested fix worked to prevent the error message! Thanks for that! Of course a new MethodError appeared in its place, this time `MethodError: no method matching floatmin(::Type{Posit{16,1}})`. However, I suppose I can just use the above fix-recipe of temporarily defining my own new methods to resolve these on a case by case basis. 

For your information @giordano,  as regards the Apple M1 issue, what happens after running `Pkg.add("SoftPosit")` and `using SoftPosit` is that it throws two errors: 

`ERROR: LoadError: SoftPosit not installed properly, run "]build SoftPosit", restart Julia and try again`

`ERROR: Failed to precompile SoftPosit [0775deef-a35f-56d7-82da-cfc52f91364d] to /Users/alexandrosstamatiou/.julia/compiled/v1.7/SoftPosit/jl_DmAp88.`

When you follow the "]build SoftPosit" approach it doesn't work:

`ERROR: Error building SoftPosit:` 
`┌ Warning: Platform arm64-apple-darwin21.2.0 is not an officially supported platform`
`└ @ BinaryProvider ~/.julia/packages/BinaryProvider/U2dKK/src/PlatformNames.jl:450`
`ERROR: LoadError: KeyError: key "unknown" not found`
`ERROR: LoadError: Failed to precompile BinaryProvider [b99e7846-7c00-51b0-8f62-c81ae34c0232] to` `/Users/alexandrosstamatiou/.julia/compiled/v1.7/BinaryProvider/jl_UyYbUV.`

**Comment by giordano** [comment]
> @giordano do you know why the precompilation of SoftPosit.jl on an apple M1 would fail although [softposit_jll](https://github.com/JuliaBinaryWrappers/SoftPosit_jll.jl) should build?

There was no release after #63

**Comment by milankl** [comment]
Haha, should have realised that 🙈! @alexandrosstamatiou can you install from #main instead of the last released version? Like so
```julia
julia>] add https://github.com/milankl/SoftPosit.jl#main
```

**Comment by alexandrosstamatiou** [comment]
I will do that now, then we can use the SoftPosit.jl package from here on

**Comment by alexandrosstamatiou** [comment]
Now it works fine and the `floatmin()` issue doesn't come up. The computation carries on for a bit before we hit a type promotion issue: 
`promotion of types Posit16 and Float64 failed to change any arguments`

**Comment by alexandrosstamatiou** [comment]
For example,

`x=promote(Posit16(1),Float64(1))`

results in the same error:

**Comment by alexandrosstamatiou** [comment]
I should probably try to use the `promote_rule` in this instance, forcing them to be of type `Posit16`

**Comment by alexandrosstamatiou** [comment]
This worked: 
`Base.promote_rule(::Type{Posit16}, ::Type{Float64}) = Posit16`

**Comment by milankl** [comment]
This is a method that is essential if someone hardcoded a line like `2.0*a`, which isn't defined unless one of the types is promoted to the other. However, I deliberately not define these methods (whether you convert to high precision as Float16*Float64 would, or down to make sure everything is done in posits for example) as you easily overlook where these promotions happen. Hence you easily end up with code that is type-flexible but only through a bunch of conversions that ideally should be avoided in the first place (or only done upfront). That's why I'd rather have an error message telling me where I wrote `2.0*a`...

**Comment by alexandrosstamatiou** [comment]
So what would you do to make `2.0*a` work, in order for the calculation to continue?

**Comment by milankl** [comment]
Write it as `convert(T,2)*a` to be explicit or if it's a constant that's used repeatedly
```julia
two = convert(T,2)
b = two*a
```
but that's also my style as I want to be as explicit as possible to avoid any unintentional conversion which can cause errors as the one above, or maybe even worse, you start with one number format and suddenly cascade into another...

**Comment by milankl** [comment]
BigFloat conversion addressed with #68
